src/chainstate/burn/db/burndb.rs

```diff
@@ -604,16 +679,252 @@ impl BurnDB {
+                   VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16, ?17, ?18, ?19)", args)
             .map_err(db_error::SqliteError)?;
 
         Ok(())
     }
 
+    /// Get the current arrival index.
+    fn get_arrival_index(conn: &Connection) -> Result<u64, db_error> {
+        match query_row::<BlockSnapshot, _>(conn, "SELECT * FROM snapshots ORDER BY arrival_index DESC LIMIT 1", NO_PARAMS)? {
+            Some(sn) => Ok(sn.arrival_index),
+            None => Ok(0)
+        }
+    }
+
+    /// MARF index key for a processed stacks block.  Maps to its height.
+    fn make_stacks_block_index_key(stacks_block_hash: &BlockHeaderHash) -> String {
+        format!("burndb::stacks::block::{}", stacks_block_hash)
+    }
+
+    /// MARF index value for a processed stacks block
+    fn make_stacks_block_index_value(height: u64) -> String {
+        format!("{}", height)
+    }
+
+    /// MARF index key for the highest arrival index processed in a fork
+    fn make_stacks_block_max_arrival_index_key() -> String {
+        "burndb::stacks::block::max_arrival_index".to_string()
+    }
+
+    /// MARF index value for the highest arrival index processed in a fork
+    fn make_stacks_block_max_arrival_index_value(index: u64) -> String {
+        format!("{}", index)
+    }
+
+    /// Add an accepted Stacks block to the canonical accepted stacks header table, to indicate
+    /// that it will be committed to by the next burn block added to the canonical chain tip.  Used
+    /// to identify Stacks blocks that get accepted in the mean time, so we can ensure that the
+    /// canonical burn chain tip always points to the canonical stacks chain tip.
+    fn insert_accepted_stacks_block_pointer<'a>(tx: &mut BurnDBTx<'a>, tip_burn_header_hash: &BurnchainHeaderHash, burn_header_hash: &BurnchainHeaderHash, stacks_block_hash: &BlockHeaderHash, stacks_block_height: u64) -> Result<(), db_error> {
+        let args: &[&dyn ToSql] = &[tip_burn_header_hash, burn_header_hash, stacks_block_hash, &u64_to_sql(stacks_block_height)?];
+        tx.execute("INSERT OR REPLACE INTO canonical_accepted_stacks_blocks (tip_burn_block_hash, burn_block_hash, stacks_block_hash, block_height) VALUES (?1, ?2, ?3, ?4)", args)
+            .map_err(db_error::SqliteError)?;
+        Ok(())
+    }
+
+    /// Get an accepted stacks block header in a fork whose chain tip has not yet committed
+    /// to it.
+    fn get_accepted_stacks_block_pointer(conn: &Connection, tip_burn_header_hash: &BurnchainHeaderHash, stacks_block_hash: &BlockHeaderHash) -> Result<Option<AcceptedStacksBlockHeader>, db_error> {
+        let args : &[&dyn ToSql] = &[tip_burn_header_hash, stacks_block_hash];
+        query_row(conn, "SELECT * FROM canonical_accepted_stacks_blocks WHERE tip_burn_block_hash = ?1 AND stacks_block_hash = ?2", args)
```

Comment:
Although you can't enforce it in Sqlite, this query is guaranteed to never return more than one row.  This is because Stacks block entries don't ever get added to `canonical_accepted_stacks_blocks` unless they are first processed by `StacksChainState::process_next_staging_block()`.  By construction, each Stacks block in a given burn chain fork (identified by `tip_burn_header_hash`) will have a unique hash within that fork.  But sure, I can add this check.

Reply:
Note that  `util::db::query_expect_row` implements a version that panics if more than one row is returned.

Reply:
Fixed in 90c7c96